certbot/certbot/_internal/plugins/selection.py

```diff
@@ -249,6 +249,13 @@ def choose_configurator_plugins(config: configuration.NamespaceConfig,
             installer = pick_installer(config, req_inst, plugins, installer_question)
         if need_auth:
             authenticator = pick_authenticator(config, req_auth, plugins)
+
+    # As a special case for certonly, if the user interactively selected nginx
+    # or apache as an authenticator, make sure we set the appropriate installer
+    # to allow us to safely reload after retrieving a cert
+    if verb == "certonly" and authenticator and not installer:
+        if authenticator.name in ("nginx", "apache"):
+            installer = cast(Optional[interfaces.Installer], authenticator)
```

Comment:
This also means that:

     certbot certonly -d example.com --authenticator nginx --installer none

results in an `installer` being set in ` /etc/letsencrypt/renewal/example.com.conf` anyway.

Do we want this to happen? Is there any difference between `certonly` pretending there's an `installer` set, and  forcefully setting an `installer` like this PR does?

If the latter, I think we will also need to reword the CHANGELOG to reflect the change being made here.

Reply:
Ah nice catch, unfortunately I do think we ideally wouldn't have that behavior, since I feel like specifying `-i none` would only be done when the user specifically does not want any installer stuff to happen at all, like renewing, and maybe in that case we shouldn't even install the cert in the first place. Unless there's some other reason I'm missing why someone would specify `-i none`?

That being said, detecting that case would require differentiating between the default value and `none` set on the cli, which last I checked is a pretty hacky bug-prone part of the codebase. Maybe it's gotten better since then or maybe it's sufficiently smoothed over by higher-level functions these days that hide the messiness, but I think my sense of what decision to make depends on the cost of bringing that code in.

Reply:
i wonder if we could get away with handling just these two cases, which i think are easier to detect from `config`:

1. User used `--nginx-`/`--apache`/`--configurator nginx/apache` in the CLI, in which case we set the installer for certonly
2. User used `--authenticator nginx/apache` but didn't set an installer, in which case we assume the did so intentionally and leave the installer blank

Reply:
Yeah, `config` is what I meant by the bug-prone area of the codebase; it's detecting case 2 vs explicitly set it to `none` that might be tricky. It might not though! Those are in fact the relevant cases, if you want to go ahead and see how icky it'll be it's at least best to know so we can decide if it's worth doing.

Reply:
ah fwiw i was considering `--authenticator nginx --installer none` and `--authenticator nginx` to both be case 2. if that's acceptable, i think this works:

```
    # As a special case for certonly, if a user selected apache or nginx, set
    # the relevant installer (unless the user specifically specified no
    # installer)
    if verb == "certonly":
        # user specified --nginx or --apache on CLI
        selected_configurator = config.nginx or config.apache
        # user didn't request an authenticator, and so interactively chose nginx
        # or apache
        interactively_selected = req_auth is None and authenticator.name in ("nginx", "apache")

        if selected_configurator or interactively_selected:
            installer = cast(Optional[interfaces.Installer], authenticator)
            ```

Reply:
Gave that a quick test and it seems to do just what we want. Nice!